portfolio/src/main/java/com/google/sps/servlets/RegionDataServlet.java

```diff
@@ -0,0 +1,72 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/region-data")
+public class RegionDataServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    Map<String, Integer> regionVotes = getVotes();
+
+    response.setContentType("application/json");
+    Gson gson = new Gson();
+    String json = gson.toJson(regionVotes);
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String region = request.getParameter("region");
+
+    Entity voteEntity = new Entity("Vote");
+    voteEntity.setProperty("name", region);
+
+    writeToDatabase(voteEntity);
+
+    response.sendRedirect("/index.html#forfun");
+  }
+
+  private void writeToDatabase(Entity toWrite) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(toWrite);
+  }
+
+  private Map<String, Integer> getVotes() {
```

Comment:
Hm, this tabulates the votes each time you need to render the page which could potentially become very costly if there's millions of votes. Instead of recording individual votes, could you fetch the current count, and then increment it in your doPost method?

If you're interested in a bit more of a challenge/learning exercise (one which is quite relevant to day to day work we do with server programming), one catch to the above is the race condition you'd get if two people try to vote at the same time:
- Suppose the count for some country is 5;
- User A votes, which would cause your servlet to read the value 5, and increment it to 6
- But before it has a chance to write the result (6) to the datastore, User B votes.
- User B still sees the old value 5 in datastore, and increments its copy to 6.
- Now A writes storing 6 in the datastore
- Now B writes, storing 6 in the datastore again

This can happen because the servlet is multithreaded. Here's a video explaining the problem and the "synchronized" term in java:
https://www.youtube.com/watch?v=RH7G-N2pa8M
If marking the doPost method as synchronized doesn't work, a synchronized block might work instead.